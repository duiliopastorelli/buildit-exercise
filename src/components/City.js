import React, {Component} from 'react';
import PropTypes from 'prop-types';
import Card from './Card';

//todo make offset and precision amendable by the user

/**
 * This component has the responsibility of handle all the forecast data.
 * It shows a "loading" message for keep updated the user about what is
 * happening behind the scenes.
 * When data are available it pass them to the children Card components.
 */
class City extends Component {
  render() {
    const {cityName, countryName, forecastedDays} = this.props;

    return (
      <div>
        {/* Displays a loading message if data are not available yet */}
        {cityName ?
          <h1>{cityName} ({countryName})</h1> :
          <p>Loading</p>}

        <div className="cards--container">
          {forecastedDays.map((day, index) => {
            //Filter out the forecasts that don't respect the offset and the
            // precision chosen
            if ((index - 2) % 8 === 0) {
              return <Card key={day.dt}
                           date={day.dt}
                           temp={day.main.temp}
                           tempMin={day.main.temp_min}
                           tempMax={day.main.temp_max}
                           weather={day.weather[0].main}
                           weatherDescription={day.weather[0].description}
                           icon={day.weather[0].icon}
                           windSpeed={day.wind.speed}
                           windDirection={day.wind.deg}
              />
            }
          })}
        </div>
      </div>
    )
  }
}

City.propTypes = {
  cityName: PropTypes.string.isRequired,
  countryName: PropTypes.string.isRequired,
  forecastedDays: PropTypes.array.isRequired
};

export default City;