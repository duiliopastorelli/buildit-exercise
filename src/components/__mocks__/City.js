export const props = {
  "cityName": "Warsawa",
  "countryName": "Poland",
  "forecastedDays": [
    {
      "dt": 1538740800,
      "main": {
        "temp": 14.23,
        "temp_min": 14.23,
        "temp_max": 17.3,
      },
      "weather": [
        {
          "id": 800,
          "main": "Clear",
          "description": "clear sky",
          "icon": "01d"
        }
      ],
      "wind": {
        "speed": 5.02,
        "deg": 232.501
      },
    },
    {
      "dt": 1538740800,
      "main": {
        "temp": 14.23,
        "temp_min": 14.23,
        "temp_max": 17.3,
      },
      "weather": [
        {
          "id": 800,
          "main": "Clear",
          "description": "clear sky",
          "icon": "01d"
        }
      ],
      "wind": {
        "speed": 5.02,
        "deg": 232.501
      },
    },
    {
      "dt": 1538740800,
      "main": {
        "temp": 14.23,
        "temp_min": 14.23,
        "temp_max": 17.3,
      },
      "weather": [
        {
          "id": 800,
          "main": "Clear",
          "description": "clear sky",
          "icon": "01d"
        }
      ],
      "wind": {
        "speed": 5.02,
        "deg": 232.501
      },
    },
    {
      "dt": 1538740800,
      "main": {
        "temp": 14.23,
        "temp_min": 14.23,
        "temp_max": 17.3,
      },
      "weather": [
        {
          "id": 800,
          "main": "Clear",
          "description": "clear sky",
          "icon": "01d"
        }
      ],
      "wind": {
        "speed": 5.02,
        "deg": 232.501
      },
    },
    {
      "dt": 1538740800,
      "main": {
        "temp": 14.23,
        "temp_min": 14.23,
        "temp_max": 17.3,
      },
      "weather": [
        {
          "id": 800,
          "main": "Clear",
          "description": "clear sky",
          "icon": "01d"
        }
      ],
      "wind": {
        "speed": 5.02,
        "deg": 232.501
      },
    },
    {
      "dt": 1538740800,
      "main": {
        "temp": 14.23,
        "temp_min": 14.23,
        "temp_max": 17.3,
      },
      "weather": [
        {
          "id": 800,
          "main": "Clear",
          "description": "clear sky",
          "icon": "01d"
        }
      ],
      "wind": {
        "speed": 5.02,
        "deg": 232.501
      },
    },
    {
      "dt": 1538740800,
      "main": {
        "temp": 14.23,
        "temp_min": 14.23,
        "temp_max": 17.3,
      },
      "weather": [
        {
          "id": 800,
          "main": "Clear",
          "description": "clear sky",
          "icon": "01d"
        }
      ],
      "wind": {
        "speed": 5.02,
        "deg": 232.501
      },
    },
    {
      "dt": 1538740800,
      "main": {
        "temp": 14.23,
        "temp_min": 14.23,
        "temp_max": 17.3,
      },
      "weather": [
        {
          "id": 800,
          "main": "Clear",
          "description": "clear sky",
          "icon": "01d"
        }
      ],
      "wind": {
        "speed": 5.02,
        "deg": 232.501
      },
    }],
  "fetchNewCity": function () {
  },
};