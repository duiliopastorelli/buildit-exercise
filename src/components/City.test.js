import React from 'react';
import City from './City';
import {configure, shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import {props} from './__mocks__/City';
import renderer from 'react-test-renderer';

configure({adapter: new Adapter()});

describe('City Component', () => {

  it('renders the City component', () => {
    const component = shallow(<City {...props} />);
    expect(component.find('CityDialog').length).toBe(1);
    expect(component.find('Card').length).toBe(1);
  });

  it('renders correctly', () => {
    const tree = renderer
      .create(<City {...props} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

});