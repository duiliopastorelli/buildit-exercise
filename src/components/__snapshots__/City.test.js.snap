// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`City Component renders correctly 1`] = `
<div>
  <h1>
    Warsawa
     (
    Poland
    )
  </h1>
  <div>
    <button
      className="MuiButtonBase-root-27 MuiButton-root-1 MuiButton-text-3 MuiButton-flat-6"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiButton-label-2"
      >
        Change city
      </span>
      <span
        className="MuiTouchRipple-root-30"
      />
    </button>
  </div>
  <div
    className="cards--container"
  >
    <div
      className="MuiPaper-root-52 MuiPaper-elevation2-56 MuiPaper-rounded-53 card"
    >
      <span
        className="card--date"
      >
        5
        /
        10
        /
        2018
         - 
        14
        :00
      </span>
      <div
        className="card--main-info"
      >
        <img
          alt="clear sky"
          src="http://openweathermap.org/img/w/01d.png"
        />
        <div
          className="card--temperatures"
        >
          <span>
            14.23
            ºC - 
            Clear
          </span>
          <span>
            14.23
            ºC ~ 
            17.3
            ºC
          </span>
        </div>
      </div>
      <span
        className="card--weather-description"
      >
        clear sky
      </span>
      <span
        className="card--wind"
      >
        Wind 
        5.02
        km/h @ 
        232.501
        º
      </span>
    </div>
  </div>
</div>
`;
